### PR TITLE
SAML/SLO: Return a response in logout endpoint

### PR DESCRIPTION
**Pull request by tung2744**
ref DEV-2060

TODO:
- [x] Put the signature of the response to query if using redirect binding

### CODE REVIEW

File: pkg/auth/handler/saml/logout.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	r *http.Request,
+	sp *config.SAMLServiceProviderConfig,
+	callbackURL string,
+) (relayState string, result samlprotocolhttp.SAMLResult) {
+	now := h.Logger.Time.UTC()
+	var parseResult samlbinding.SAMLBindingParseResult
+	var logoutRequest *samlprotocol.LogoutRequest
+
+	// Get data with corresponding binding
+	var err error
```
Comment (louischan-oursky) [inline pkg/auth/handler/saml/logout.go:79]: 1. Prefer more than one methods returning (A, error). The error could be a error result that can be written directly.
2. One method to see the flow e.g. handleSLORequest

```
func handleSLORequest() {
  request, err := parse(r)
}
```


```
var err error

defer func() {
  if err != nil {
    // ...
  }
}()



switch {
case A
  if err != nil {
    return
  }
case B
}
```

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
1. Delete samlerror. Move the contents to samlprotocol/error_blahblahblah.go
2. Delete samlprotocolhttp. Move the contents to samlbinding.
3. Remove IsInternalError()

**Comment by tung2744** [comment]
Updated, thanks!

File: pkg/auth/handler/saml/logout.go [hunk 2]
```diff
@@ -147,28 +164,34 @@ func (h *LogoutHandler) handleSLORequest(
 			break
```
Comment (louischan-oursky) [inline pkg/auth/handler/saml/logout.go:164]: No break

File: pkg/auth/handler/saml/logout.go [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+		if parseErr != nil {
+			return nil, nil, "", parseErr
+		}
+		logoutRequest, err = samlprotocol.ParseLogoutRequest([]byte(r.SAMLRequestXML))
+		if err != nil {
+			err = &samlprotocol.ParseRequestFailedError{
+				Reason: "malformed LogoutRequest",
+				Cause:  err,
+			}
+			break
```
Comment (louischan-oursky) [inline pkg/auth/handler/saml/logout.go:109]: No break

File: pkg/auth/handler/saml/logout.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
```
Comment (louischan-oursky) [inline pkg/auth/handler/saml/logout.go:125]: No break

File: pkg/auth/handler/saml/logout.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+			parseResult.SigAlg,
+			parseResult.RelayState,
+			parseResult.Signature)
+		if err != nil {
+			break
+		}
+	case *samlbinding.SAMLBindingHTTPPostParseResult:
+		err = h.SAMLService.VerifyEmbeddedSignature(sp, parseResult.SAMLRequestXML)
+		if err != nil {
+			break
```
Comment (louischan-oursky) [inline pkg/auth/handler/saml/logout.go:169]: No break

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Review by louischan-oursky — APPROVED** [review]

**Comment by tung2744** [comment]
I found that I forgot to remove the samlerror package, although no code is using them. I will remove them in next pr.

**Pull request merged**